baseplate/queue_consumer.py

```diff
@@ -0,0 +1,201 @@
+    :param handler: The handler method.
+
+    """
+    queues = []
+    for routing_key in routing_keys:
+        queue = Queue(
+            name=queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+        )
+        queues.append(queue)
+
+    logger.info("registering %s as a handler for %r", handler.__name__, queues)
+    kombu_consumer = KombuConsumer.new(connection, queues)
+
+    logger.info("waiting for messages")
+    while True:
+        context = ConsumerContext()
+        with baseplate.make_server_span(context, queue_name) as span:
+            message = kombu_consumer.get_message(span)
+            handler(context, message.body, message)
+            message.ack()
+
+
+class _ConsumerWorker(ConsumerMixin):
+    def __init__(self, connection, queues, work_queue):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+
+    def get_consumers(self, Consumer, channel):
+        return [Consumer(
+            queues=self.queues,
+            on_message=self.on_message,
+        )]
+
+    def on_message(self, message):
+        self.work_queue.put(message)
+
+    def get_message(self, block, timeout):
+        try:
+            return self.work_queue.get(block=block, timeout=timeout)
+        except queue.Empty:
+            return None
+
+class ConsumerContext(object):
+    pass
+
+
+class BaseKombuConsumer(object):
+    """Base object for consuming messages from a queue.
+
+    A worker process accepts messages from the queue and puts them in a local
+    work queue. The "real" consumer can then get messages with
+    :py:meth:`get_message` or :py:meth:`get_batch`. It is that consumer's
+    responsibility to ack or reject messages.
+
+    Can be used directly, outside of standard baseplate context.
+
+    """
+    def __init__(self, worker, worker_thread):
+        self.worker = worker
+        self.worker_thread = worker_thread
+
+    @classmethod
+    def new(cls, connection, queues):
+        work_queue = queue.Queue()
+        worker = _ConsumerWorker(connection, queues, work_queue)
+        worker_thread = Thread(target=worker.run)
+        worker_thread.name = "consumer message pump"
+        worker_thread.daemon = True
+        worker_thread.start()
+
+        return cls(worker, worker_thread)
+
+    def get_message(self):
+        """Return a single message."""
+        batch = self.get_batch(max_items=1, timeout=None)
+        return batch[0]
+
+    def get_batch(self, max_items, timeout):
+        """Return a batch of messages.
+
+        :param int max_items: The maximum batch size.
+        :param int timeout: The maximum time to wait in seconds, or ``None``
+            for no timeout.
+
+        """
+        if timeout == 0:
+            block = False
+        else:
+            block = True
+
+        batch = []
+        retry_policy = retry.RetryPolicy.new(attempts=max_items, budget=timeout)
```

Comment:
Classes are generally imported directly in baseplate (bare functions are kept with their modules)

Reply:
I had to do this to get the tests to play nice, but I'll just change my mock